From d462daa21164a52091dcf4ce069775d1a0b5d617 Mon Sep 17 00:00:00 2001
From: Jordan Lazare <Jordan.Lazare@amd.com>
Date: Tue, 4 Oct 2016 14:38:05 -0400
Subject: [PATCH 1176/1722] drm/amd/dal: Fix segfault due to null TG pointer
 with edp

On Linux, core_stream is null but we still want to set backlight level

Change-Id: I552632a2f5c2b0177567c7a550d128dce1e8f72c
Signed-off-by: Jordan Lazare <Jordan.Lazare@amd.com>
Acked-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/dc/core/dc_link.c | 29 +++++++++++++++--------------
 1 file changed, 15 insertions(+), 14 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc_link.c b/drivers/gpu/drm/amd/dal/dc/core/dc_link.c
index 61ba456..d9a67ca 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc_link.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc_link.c
@@ -1398,16 +1398,6 @@ bool dc_link_set_backlight_level(const struct dc_link *dc_link, uint32_t level,
 	int i;
 	uint32_t dmcu_status;
 
-	for (i = 0; i < MAX_PIPES; i++) {
-		if (core_dc->current_context->res_ctx.pipe_ctx[i].stream
-				== core_stream)
-			/* dmcu -1 for all controller id values,
-			 * therefore +1 here
-			 */
-			controller_id = core_dc->current_context->res_ctx.
-					pipe_ctx[i].tg->inst + 1;
-	}
-
 	dal_logger_write(ctx->logger, LOG_MAJOR_BACKLIGHT,
 			LOG_MINOR_BACKLIGHT_INTERFACE,
 			"New Backlight level: %d (0x%X)\n", level, level);
@@ -1415,13 +1405,24 @@ bool dc_link_set_backlight_level(const struct dc_link *dc_link, uint32_t level,
 	dmcu_status = dm_read_reg(ctx, mmDMCU_STATUS);
 
 	/* If DMCU is in reset state, DMCU is uninitialized */
-	if (get_reg_field_value(dmcu_status, mmDMCU_STATUS, UC_IN_RESET))
+	if (get_reg_field_value(dmcu_status, mmDMCU_STATUS, UC_IN_RESET)) {
 		link->link_enc->funcs->set_lcd_backlight_level(link->link_enc,
 						level);
-	else
+	} else {
+		for (i = 0; i < MAX_PIPES; i++) {
+			if (core_dc->current_context->res_ctx.pipe_ctx[i].stream
+					== core_stream)
+				/* dmcu -1 for all controller id values,
+				 * therefore +1 here
+				 */
+				controller_id = core_dc->current_context->res_ctx.
+						pipe_ctx[i].tg->inst + 1;
+		}
+
 		link->link_enc->funcs->set_dmcu_backlight_level
-					(link->link_enc, level,
-					frame_ramp, controller_id);
+				(link->link_enc, level,
+				frame_ramp, controller_id);
+	}
 	return true;
 }
 
-- 
2.7.4

